SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracer.cpp

```diff
@@ -0,0 +1,226 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "Interop/Connection/SpatialEventTracer.h"
+
+#include "EngineClasses/SpatialNetDriver.h"
+#include "EngineMinimal.h"
+#include "GameFramework/Actor.h"
+#include "GenericPlatform/GenericPlatformAtomics.h"
+#include "Interop/Connection/SpatialConnectionManager.h"
+#include "Interop/Connection/SpatialWorkerConnection.h"
+#include "UObject/Class.h"
+#include <WorkerSDK/improbable/c_trace.h>
+#include <WorkerSDK/improbable/c_io.h>
+
+DEFINE_LOG_CATEGORY(LogSpatialEventTracer);
+
+using namespace SpatialGDK;
+using namespace worker::c;
+
+void SpatialEventTracer::TraceCallback(void* UserData, const Trace_Item* Item)
+{
+	SpatialEventTracer* EventTracer = static_cast<SpatialEventTracer*>(UserData);
+	checkf(EventTracer->bEnabled && EventTracer->Stream, TEXT("Tracer in an invalid state in TraceCallback")); 
+	if (EventTracer->BytesWrittenToStream < EventTracer->MaxFileSize)
+	{
+		uint32_t ItemSize = Trace_GetSerializedItemSize(Item); 
+		EventTracer->BytesWrittenToStream += ItemSize;
+		int Code = Trace_SerializeItemToStream(EventTracer->Stream, Item, ItemSize);
+		if (Code != 0)
+		{
+			UE_LOG(LogSpatialEventTracer, Error, TEXT("Failed to serialize to with error code %d (%s"), Code, Trace_GetLastError());
+		}
+	}
+}
+
+SpatialSpanIdActivator::SpatialSpanIdActivator(SpatialEventTracer* InEventTracer, const TOptional<Trace_SpanId>& InCurrentSpanId)
+	: CurrentSpanId(InCurrentSpanId)
+	, EventTracer(InEventTracer->GetWorkerEventTracer())
+{
+	if (InCurrentSpanId.IsSet())
+	{
+		Trace_EventTracer_SetActiveSpanId(EventTracer, CurrentSpanId.GetValue());
+	}
+}
+
+SpatialSpanIdActivator::~SpatialSpanIdActivator()
+{
+	if (CurrentSpanId.IsSet())
+	{
+		Trace_EventTracer_ClearActiveSpanId(EventTracer);
+	}
+}
+
+SpatialEventTracer::SpatialEventTracer(const FString& WorkerName)
+{
+	if (const USpatialGDKSettings* Settings = GetDefault<USpatialGDKSettings>())
+	{
+		if (Settings->bEventTracingEnabled)
+		{
+			MaxFileSize = Settings->MaxEventTracingFileSizeBytes;
+			Enable(WorkerName);
+		}
+	}
+}
+
+SpatialEventTracer::~SpatialEventTracer()
+{
+	if (IsEnabled())
+	{
+		Disable();
+	}
+}
+
+Trace_SpanId SpatialEventTracer::CreateNewSpanId()
+{
+	return Trace_EventTracer_AddSpan(EventTracer, nullptr, 0);
+}
+
+Trace_SpanId SpatialEventTracer::CreateNewSpanId(const TArray<Trace_SpanId>& Causes)
+{
+	return Trace_EventTracer_AddSpan(EventTracer, Causes.GetData(), Causes.Num());
+}
+
+TOptional<Trace_SpanId> SpatialEventTracer::TraceEvent(const FEventMessage& EventMessage, const UStruct* Struct, const worker::c::Trace_SpanId* Cause)
```

Comment:
What sort of thing can be passed as `Struct` here? Ideally the event data produced by a given event type (e.g. `unreal_gdk.send_rpc`) should be relatively stable (e.g. if, in one version, such an event has an `id` field, that ID field should not be removed in the next version, and if it is, it should ideally be highlighted as a breaking change).

This is so that people can derive meaning from/build analytical tooling on top of such event data without fear of it breaking underneath them.

Reply:
At the moment only subclasses of FEventMessage. This means all event message will have a string "type" field. Other than that, all data is add per derived FEventMessage. Currently it is possible for someone to remove the "Function" property from the derived FEventSendRPC message.